src/dynamics/mpas/dyn_grid.F90

```diff
@@ -1,13 +1,476 @@
+
+        ! In MPAS, zeta coordinates are stored in increasing order (i.e., bottom to top of atmosphere).
+        ! In CAM-SIMA, however, index order is reversed (i.e., top to bottom of atmosphere).
+        ! Compute in reverse below.
+        zw(pverp) = 0.0_kind_r8
+
+        do k = pver, 1, -1
+            zw(k) = zw(k + 1) + dzw(pver - k + 1)
+            zu(k) = 0.5_kind_r8 * (zw(k + 1) + zw(k))
+        end do
+
+        ! Compute reference pressure from reference height.
+        allocate(p_ref_int(pverp), stat=ierr)
+        call check_allocate(ierr, 'init_reference_pressure', 'p_ref_int', 'dyn_grid', __LINE__)
+
+        call std_atm_pres(zw, p_ref_int)
+
+        allocate(p_ref_mid(pver), stat=ierr)
+        call check_allocate(ierr, 'init_reference_pressure', 'p_ref_mid', 'dyn_grid', __LINE__)
+
+        p_ref_mid(:) = 0.5_kind_r8 * (p_ref_int(1:pver) + p_ref_int(2:pverp))
+
+        ! Print a nice table of reference height and pressure.
+        call dyn_debug_print('Reference layer information:')
+        call dyn_debug_print('----- | -------------- | --------------')
+        call dyn_debug_print('Index |     Height (m) | Pressure (hPa)')
+
+        do k = 1, pver
+            call dyn_debug_print(repeat('-', 5) // ' |  ' // stringify([zw(k)]) // &
+                ' |  ' // stringify([p_ref_int(k) / 100.0_kind_r8]))
+
+            l = len(stringify([k]))
+
+            call dyn_debug_print(repeat(' ', 5 - l) // stringify([k]) // ' |  ' // stringify([zu(k)]) // &
+                ' |  ' // stringify([p_ref_mid(k) / 100.0_kind_r8]))
+        end do
+
+        k = pverp
+
+        call dyn_debug_print(repeat('-', 5) // ' |  ' // stringify([zw(k)]) // &
+            ' |  ' // stringify([p_ref_int(k) / 100.0_kind_r8]))
+
+        call ref_pres_init(p_ref_int, p_ref_mid, num_pure_p_lev)
+    end subroutine init_reference_pressure
+
+    !> Initialize physics grid in terms of dynamics decomposition.
+    !> Provide grid and mapping information between global and local indexes to physics by calling `phys_grid_init`.
+    !> (KCW, 2024-03-27)
+    subroutine init_physics_grid()
+        character(max_hcoordname_len), allocatable :: dyn_attribute_name(:)
+        integer :: hdim1_d, hdim2_d
+        integer :: i
+        integer :: ierr
+        integer, allocatable :: indextocellid(:)  ! Global indexes of cell centers.
+        real(kind_r8), allocatable :: areacell(:) ! Cell areas (square meters).
+        real(kind_r8), allocatable :: latcell(:)  ! Cell center latitudes (radians).
+        real(kind_r8), allocatable :: loncell(:)  ! Cell center longitudes (radians).
+        type(physics_column_t), allocatable :: dyn_column(:) ! Grid and mapping information between global and local indexes.
+
+        hdim1_d = ncells_global
+
+        ! Setting `hdim2_d` to `1` indicates unstructured grid.
+        hdim2_d = 1
+
+        call mpas_dynamical_core % get_variable_value(areacell, 'mesh', 'areaCell')
```

Comment:
Here and other places, could you explain the reason for using `get_variable_value` rather than `get_variable_pointer`. Wouldn't using `get_variable_value` unnecessarily create a copy of, e.g., `areaCell`?

Reply:
Yes, it would, but the rationale behind this can be best explained by my code comments here: https://github.com/ESCOMP/CAM-SIMA/blob/e47f906b847d5d158fc290db2a924fffe075ba26/src/dynamics/mpas/driver/dyn_mpas_subdriver.F90#L1662-L1685

One of the benefits of adopting OOP design for MPAS dycore is that there is a natural barrier between the instance's internals and the outside world. This barrier ensures MPAS dycore's memory integrity. The `get_variable_pointer` subroutine family breaks this barrier, whereas it is totally safe to use the `get_variable_value` subroutine family.

However, I do recognize that exceptional circumstances demand exceptional measures. Soon MPAS dycore will finish the initialization phase and start sharing state updates with the rest of CAM-SIMA. State variables occupy large amount of memory and it is likely that I will be forced to take the "pointer frenzy" route. This is why the `get_variable_pointer` subroutine family exists in the first place.

Let me know what you think. I am fine changing them to pointers.

Reply:
I appreciate the point about ensuring the dycore's memory integrity by only providing copies of fields to the rest of CAM. On balance, though, my inclination would be to avoid making extra copies of arrays to keep CAM-SIMA as lightweight as possible with regard to memory usage. If other parts of CAM make a mess of fields "owned" by the MPAS-A dycore, tracking down and fixing those issues is a one-time expense, while we would pay the price of extra memory usage (translating to a larger number of required nodes for a simulation, and possibly less cache reuse) perpetually.

Reply:
I pushed https://github.com/ESCOMP/CAM-SIMA/pull/257/commits/f99ae47a9da22b6356b60a7ddb0b61b5ee516d79 to replace allocatables with pointers. Compile and run-tested.

Reply:
Thanks!